MSGraphCoreSDK/MSGraphCoreSDK/Middleware/Implementations/HTTPProvider/MSURLSessionManager.m

```diff
@@ -0,0 +1,253 @@
+    }
+}
+
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location
+{
+    MSURLSessionTaskDelegate *delegate = [self getDelegateForTask:downloadTask];
+    
+    if (delegate)
+    {
+        [delegate task:downloadTask didCompleteDownload:location];
+        [delegate task:downloadTask didCompleteWithError:nil];
+        // remove the task now so we don't call the completion handler when the completion delegate method gets called
+        [self removeTaskDelegateForTask:downloadTask];
+    }
+}
+
+- (void)URLSession:(NSURLSession *)session
+              task:(NSURLSessionTask *)task
+willPerformHTTPRedirection:(NSHTTPURLResponse *)redirectResponse
+        newRequest:(NSURLRequest *)request
+ completionHandler:(void (^)(NSURLRequest *))completionHandler
+{
+    MSURLSessionTaskDelegate *delegate = [self getDelegateForTask:task];
+
+    if (delegate){
+        [delegate task:task didRedirectWithResponse:redirectResponse];
+    }
+    [self removeTaskDelegateForTask:task];
+}
+
+
+#pragma mark - Middelware methods
+- (void)execute:(MSURLSessionTask *)task withCompletionHandler:(HTTPRequestCompletionHandler)completionHandler
+{
+    [task setSDKVersionRequestHeader];
+    NSURLSessionTask *sessionTask;
+    if([task isKindOfClass:[MSURLSessionDataTask class]])
+    {
+        sessionTask = [self dataTaskWithRequest:task.request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+            completionHandler(data, response, error);
+        }];
+    }else if([task isKindOfClass:[MSURLSessionDownloadTask class]])
+    {
+        NSProgress *progress = [(MSURLSessionDownloadTask *)task progress];
+        sessionTask = [self downloadTaskWithRequest:task.request progress:&progress completionHandler:^(NSURL *location, NSURLResponse *response, NSError *error) {
+            completionHandler(location, response, error);
+        }];
+    }
+    else if([task isKindOfClass:[MSURLSessionUploadTask class]])
+    {
+        NSProgress *progress = [(MSURLSessionUploadTask *)task progress];
+        if([(MSURLSessionUploadTask *)task isFileUploadTask])
+        {
+            sessionTask = [self uploadTaskWithRequest:task.request fromFile:[(MSURLSessionUploadTask *)task fileURL] progress:&progress completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+                completionHandler(data, response, error);
+            }];
+        }
+        else
+        {
+            sessionTask = [self uploadTaskWithRequest:task.request fromData:[(MSURLSessionUploadTask *)task data] progress:&progress completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+                completionHandler(data, response, error);
+            }];
+        }
+    }
+    [task setInnerTask:sessionTask];
+    [sessionTask resume];
+}
+
+- (void)setNext:(id<MSGraphMiddleware>)nextMiddleware
+{
+    if(_nextMiddleware)
```

Comment:
As this middleware is always going to be the last middleware, don't provide mechanism to set next middleware.
Remove this implementation?

Reply:
Yes, stub should be fine. Maybe with some assert warning that setnext is not allowed in this middleware.